### ISSUE TITLE
How to use Python 3.11?

### ISSUE THREAD

**Issue by vHanda** [cachix/devenv#485]
I want to use Python 3.11 for a project.

I've enabled the python language support but that gives me python 3.10. I've tried adding it to the packages via

```nix
{ pkgs, ... }:

{
  packages = [ pkgs.git  pkgs.python311 ];
  
  languages.python.enable = true;
  languages.python.poetry.enable = true;
}
```

But `python3 --version` still gives me version 3.10. I additionally tried running `nix-shell -p python311`, which then gives me python 3.11 in the shell, but poetry still uses version 3.10. I then managed to make poetry use that python version by running -

```sh
 which python3
/nix/store/9fa9qmx4bv12d58m7s26b1mnw6643f3w-python3-3.11.1/bin/python3

[nix-shell:~/src/finances]$ poetry env use /nix/store/9fa9qmx4bv12d58m7s26b1mnw6643f3w-python3-3.11.1/bin/python3
```

Is there any easier way?

**Comment by Paul75** [comment]
hello,

and adding this : 

`languages.python.package = pkgs.python311;`

?

**Comment by vHanda** [comment]
It keeps getting stuck here -

```
[1/0/7 built, 0.0 MiB DL] building python3.11-cherrypy-18.8.0 (buildPhase):   warnings.warn(msg, wardirenv: ([/opt/homebrew/Cellar/direnv/2.32.2/bin/direnv export fish]) is taking a while to execute. Use CTRL-C to give up.
[1/0/7 built, 0.0 MiB DL] building python3.11-cherrypy-18.8.0 (pytestCheckPhase): cherrypy/test/test
```

```
[1/0/7 built] building python3.11-cherrypy-18.8.0 (fixupPhase): stripping (with command strip and fldirenv: ([/opt/homebrew/Cellar/direnv/2.32.2/bin/direnv export fish]) is taking a while to execute. Use CTRL-C to give up.
error: builder for '/nix/store/1hjxbcmxba0w3q7fybfj0fq968ixqdi3-python3.11-cherrypy-18.8.0.drv' failed with exit code 1;
       last 10 log lines:
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_no_content_length
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_post_filename_with_special_characters
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_post_multipart
       > FAILED cherrypy/test/test_iterator.py::IteratorTest::test_iterator
       > FAILED cherrypy/test/test_request_obj.py::RequestObjectTests::testErrorHandling
       > FAILED cherrypy/test/test_session.py::SessionTest::test_1_Ram_Concurrency
       > FAILED cherrypy/test/test_session.py::SessionTest::test_2_File_Concurrency
       > FAILED cherrypy/test/test_tools.py::ToolTests::testHookErrors
       > = 12 failed, 260 passed, 9 skipped, 13 deselected, 2 errors in 301.61s (0:05:01) =
       > /nix/store/qrz2mnb2gsnzmw2pqax693daxh5hsgap-stdenv-darwin/setup: line 1601: pop_var_context: head of shell_variables not a function context
```

**Comment by shyim** [comment]
Can you attach your devenv.{nix,lock} and which OS?

**Comment by vHanda** [comment]
The lock file -

```
{
  "nodes": {
    "devenv": {
      "locked": {
        "dir": "src/modules",
        "lastModified": 1678703653,
        "narHash": "sha256-pUJwMOXAv1boXQW97VtZgoVc/tIU5O0nf6k7ivFsbMM=",
        "owner": "cachix",
        "repo": "devenv",
        "rev": "2ef261dcdf5cb025d0fcb0275a34228cfceb7228",
        "type": "github"
      },
      "original": {
        "dir": "src/modules",
        "owner": "cachix",
        "repo": "devenv",
        "type": "github"
      }
    },
    "flake-compat": {
      "flake": false,
      "locked": {
        "lastModified": 1673956053,
        "narHash": "sha256-4gtG9iQuiKITOjNQQeQIpoIB6b16fm+504Ch3sNKLd8=",
        "owner": "edolstra",
        "repo": "flake-compat",
        "rev": "35bb57c0c8d8b62bbfd284272c928ceb64ddbde9",
        "type": "github"
      },
      "original": {
        "owner": "edolstra",
        "repo": "flake-compat",
        "type": "github"
      }
    },
    "flake-utils": {
      "locked": {
        "lastModified": 1667395993,
        "narHash": "sha256-nuEHfE/LcWyuSWnS8t12N1wc105Qtau+/OdUAjtQ0rA=",
        "owner": "numtide",
        "repo": "flake-utils",
        "rev": "5aed5285a952e0b949eb3ba02c12fa4fcfef535f",
        "type": "github"
      },
      "original": {
        "owner": "numtide",
        "repo": "flake-utils",
        "type": "github"
      }
    },
    "gitignore": {
      "inputs": {
        "nixpkgs": [
          "pre-commit-hooks",
          "nixpkgs"
        ]
      },
      "locked": {
        "lastModified": 1660459072,
        "narHash": "sha256-8DFJjXG8zqoONA1vXtgeKXy68KdJL5UaXR8NtVMUbx8=",
        "owner": "hercules-ci",
        "repo": "gitignore.nix",
        "rev": "a20de23b925fd8264fd7fad6454652e142fd7f73",
        "type": "github"
      },
      "original": {
        "owner": "hercules-ci",
        "repo": "gitignore.nix",
        "type": "github"
      }
    },
    "nixpkgs": {
      "locked": {
        "lastModified": 1678631509,
        "narHash": "sha256-+eyd+CJUAZbk0KqqhyJtai+BmiM2/K54xE91fI2eGD0=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "b38db2c901b30a37d345751f4f6418d416e7e46e",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "nixpkgs-stable": {
      "locked": {
        "lastModified": 1673800717,
        "narHash": "sha256-SFHraUqLSu5cC6IxTprex/nTsI81ZQAtDvlBvGDWfnA=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "2f9fd351ec37f5d479556cd48be4ca340da59b8f",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixos-22.11",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "pre-commit-hooks": {
      "inputs": {
        "flake-compat": "flake-compat",
        "flake-utils": "flake-utils",
        "gitignore": "gitignore",
        "nixpkgs": [
          "nixpkgs"
        ],
        "nixpkgs-stable": "nixpkgs-stable"
      },
      "locked": {
        "lastModified": 1678376203,
        "narHash": "sha256-3tyYGyC8h7fBwncLZy5nCUjTJPrHbmNwp47LlNLOHSM=",
        "owner": "cachix",
        "repo": "pre-commit-hooks.nix",
        "rev": "1a20b9708962096ec2481eeb2ddca29ed747770a",
        "type": "github"
      },
      "original": {
        "owner": "cachix",
        "repo": "pre-commit-hooks.nix",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "devenv": "devenv",
        "nixpkgs": "nixpkgs",
        "pre-commit-hooks": "pre-commit-hooks"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

The nix file -
```
{ pkgs, ... }:

{
  packages = [ pkgs.git pkgs.jq pkgs.nodePackages.prettier ];
  languages.python.enable = true;
  languages.python.poetry.enable = true;
}
```

I'm on OSX arm64 - A Macbook Air M2.

**Comment by benaduggan** [comment]
I'm running into this issue as well on an M1 pro. Did you find any work around @vHanda?

**Comment by ziyunli** [comment]
I ran into the same issue. Apple M1 Pro w/ Ventura 13.2.1

Error:
```
error: builder for '/nix/store/b0hwqhr1jsldh35932w65ws54171j4h5-python3.11-cherrypy-18.8.0.drv' failed with exit code 1;
       last 10 log lines:
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_no_content_length
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_post_filename_with_special_characters
       > FAILED cherrypy/test/test_http.py::HTTPTests::test_post_multipart
       > FAILED cherrypy/test/test_iterator.py::IteratorTest::test_iterator
       > FAILED cherrypy/test/test_request_obj.py::RequestObjectTests::testErrorHandling
       > FAILED cherrypy/test/test_session.py::SessionTest::test_1_Ram_Concurrency
       > FAILED cherrypy/test/test_session.py::SessionTest::test_2_File_Concurrency
       > FAILED cherrypy/test/test_tools.py::ToolTests::testHookErrors
       > = 12 failed, 260 passed, 9 skipped, 13 deselected, 2 errors in 10878.95s (3:01:18) =
       > /nix/store/5pq0sy4haiy4rb2chya6qq72w95ci2dp-stdenv-darwin/setup: line 1602: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/b0hwqhr1jsldh35932w65ws54171j4h5-python3.11-cherrypy-18.8.0.drv'.
error: 1 dependencies of derivation '/nix/store/alddm2566zl717sbm476pr3qdyyrlps9-python3.11-cachecontrol-0.12.11.drv' failed to build
error: 1 dependencies of derivation '/nix/store/wrjkjqgavwx91rw1sk648g4ygrmrwpk7-python3.11-poetry-1.4.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/r302xwrblkm28gffzxx61kh998hmw55s-devenv-profile.drv' failed to build
error:
       error: 1 dependencies of derivation '/nix/store/9s01il9h4hxd33ggimh72gsgghsg348q-devenv-shell-env.drv' failed to build
direnv: error exit status 1
```

`devenv.nix`:

```nix
{ pkgs, ... }:

{
  packages = [
    pkgs.git
    pkgs.jq
    pkgs.zlib
  ];

  languages.python = {
    enable = true;
    poetry.enable = true;
    package = pkgs.python311;
  };
}
```

`devenv.lock`

```nix
{
  "nodes": {
    "devenv": {
      "locked": {
        "dir": "src/modules",
        "lastModified": 1680271030,
        "narHash": "sha256-w1Y/S3TylKylTHCp+YkOJbiAE2LBBF1sKIKuLDW4fkU=",
        "owner": "cachix",
        "repo": "devenv",
        "rev": "52232ba41b7c531eacec4cf192d87241bef1a0d1",
        "type": "github"
      },
      "original": {
        "dir": "src/modules",
        "owner": "cachix",
        "repo": "devenv",
        "type": "github"
      }
    },
    "flake-compat": {
      "flake": false,
      "locked": {
        "lastModified": 1673956053,
        "narHash": "sha256-4gtG9iQuiKITOjNQQeQIpoIB6b16fm+504Ch3sNKLd8=",
        "owner": "edolstra",
        "repo": "flake-compat",
        "rev": "35bb57c0c8d8b62bbfd284272c928ceb64ddbde9",
        "type": "github"
      },
      "original": {
        "owner": "edolstra",
        "repo": "flake-compat",
        "type": "github"
      }
    },
    "flake-utils": {
      "locked": {
        "lastModified": 1667395993,
        "narHash": "sha256-nuEHfE/LcWyuSWnS8t12N1wc105Qtau+/OdUAjtQ0rA=",
        "owner": "numtide",
        "repo": "flake-utils",
        "rev": "5aed5285a952e0b949eb3ba02c12fa4fcfef535f",
        "type": "github"
      },
      "original": {
        "owner": "numtide",
        "repo": "flake-utils",
        "type": "github"
      }
    },
    "gitignore": {
      "inputs": {
        "nixpkgs": [
          "pre-commit-hooks",
          "nixpkgs"
        ]
      },
      "locked": {
        "lastModified": 1660459072,
        "narHash": "sha256-8DFJjXG8zqoONA1vXtgeKXy68KdJL5UaXR8NtVMUbx8=",
        "owner": "hercules-ci",
        "repo": "gitignore.nix",
        "rev": "a20de23b925fd8264fd7fad6454652e142fd7f73",
        "type": "github"
      },
      "original": {
        "owner": "hercules-ci",
        "repo": "gitignore.nix",
        "type": "github"
      }
    },
    "nixpkgs": {
      "locked": {
        "lastModified": 1680273054,
        "narHash": "sha256-Bs6/5LpvYp379qVqGt9mXxxx9GSE789k3oFc+OAL07M=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "3364b5b117f65fe1ce65a3cdd5612a078a3b31e3",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "nixpkgs-stable": {
      "locked": {
        "lastModified": 1678872516,
        "narHash": "sha256-/E1YwtMtFAu2KUQKV/1+KFuReYPANM2Rzehk84VxVoc=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "9b8e5abb18324c7fe9f07cb100c3cd4a29cda8b8",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixos-22.11",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "pre-commit-hooks": {
      "inputs": {
        "flake-compat": "flake-compat",
        "flake-utils": "flake-utils",
        "gitignore": "gitignore",
        "nixpkgs": [
          "nixpkgs"
        ],
        "nixpkgs-stable": "nixpkgs-stable"
      },
      "locked": {
        "lastModified": 1680170909,
        "narHash": "sha256-FtKU/edv1jFRr/KwUxWTYWXEyj9g8GBrHntC2o8oFI8=",
        "owner": "cachix",
        "repo": "pre-commit-hooks.nix",
        "rev": "29dbe1efaa91c3a415d8b45d62d48325a4748816",
        "type": "github"
      },
      "original": {
        "owner": "cachix",
        "repo": "pre-commit-hooks.nix",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "devenv": "devenv",
        "nixpkgs": "nixpkgs",
        "pre-commit-hooks": "pre-commit-hooks"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

**Comment by vHanda** [comment]
> I'm running into this issue as well on an M1 pro. Did you find any work around @vHanda?

I did not. For this project, it was fine to downgrade to Python 3.10 since it's a pet project.

**Comment by workflow** [comment]
Using both Python 3.11 and Poetry from nixpkgs-22.11 seems to work.

`devenv.nix`

```
{ inputs, pkgs, ... }:

let
  pkgs-stable = import inputs.nixpkgs-stable { system = pkgs.stdenv.system; };

in
{
  # https://devenv.sh/basics/
  env.GREET = "Welcome to the Vest development environment!";

  # https://devenv.sh/languages/
  languages.python = {
    enable = true;
    package = pkgs-stable.python311;
    poetry = {
      enable = true;
      package = pkgs-stable.poetry.override
        {
          python = pkgs-stable.python311;
        };
    };
  };
...
```

`devenv.yaml`
```
inputs:
  nixpkgs:
    url: github:NixOS/nixpkgs/nixpkgs-unstable
  nixpkgs-stable:
    url: github:NixOS/nixpkgs/nixos-22.11
```

**Comment by benaduggan** [comment]
@workflow Worked perfectly, and also makes so much sense! Thanks!

**Comment by bobvanderlinden** [comment]
I just ran into this issue and had seen this before as well. I think the problem is that I opted to override the python version of poetry to be `config.languages.python.package`, but the version of poetry that is in nixpkgs isn't always compatible with a different python version.

I've now opted to use `pkgs.poetry` as-is in https://github.com/cachix/devenv/pull/522. That means that poetry itself will potentially run under a different version of python compared to the python package you'd like to use for your application. This should be fine, since poetry already does this when using its virtualenvs.

In versions before https://github.com/cachix/devenv/pull/522, the workaround that worked for me was to set `languages.python.poetry.package = pkgs.poetry` (a non-overridden version of poetry).
